src/coordinator/coordinator_server.cc

```diff
@@ -0,0 +1,441 @@
+#include "src/coordinator/coordinator_server.h"
+
+#include <future>
+#include <sstream>
+#include <string>
+
+#include "absl/status/statusor.h"
+#include "absl/strings/str_cat.h"
+#include "grpc/grpc.h"
+#include "grpcpp/create_channel.h"
+#include "grpcpp/security/credentials.h"
+#include "grpcpp/server_context.h"
+#include "openssl/sha.h"
+#include "src/blockchain/two_phase_commit.h"
+#include "src/proto/cohort.grpc.pb.h"
+#include "src/proto/common.pb.h"
+#include "src/proto/coordinator.grpc.pb.h"
+#include "src/utils/grpc_utils.h"
+
+namespace coordinator {
+
+namespace {
+
+using Blockchain = ::blockchain::TwoPhaseCommit;
+using CohortStub = ::cohort::Cohort::StubInterface;
+using ::common::Namespace;
+using ::grpc::ClientContext;
+using ::grpc::ServerContext;
+
+enum TransactionType {
```

Comment:
I'm unsure how much the coordinator should look into a transaction content other than the namespace. It seems reasonable to me that the coordinator can hand everything to the cohort and let them deal with the details, so coordinator don't have to worry about it at the caller level.

Reply:
To support get_cohort_namespaces, the coordinator can make trust cohort and take whatever they return without tracking the status. With this, the coordinator is much lightweight and easier to reasonable about recovery strategy etc when it's needed.

Reply:
Can you clarify what a trust cohort is? Are you suggesting that the coordinator can just ask all the cohorts for responses (i.e. not distinguish between cohorts that have read only responses)? That would simplify the logic a little.

Reply:
The only concern is that the DB interface has different methods for starting read/write vs readonly transactions. I guess the cohort could loop through the operations first to figure this out and then loop through them to process them. @n-triple-a as an FYI.

Reply:
IIUC, what Heron means is: Coordinator does not categorize the the transaction but fully trust that each cohort can pick the right transaction on its own. An example might be:
OriginalTxn {Get('Cohort1-Key1', 1), Put('Cohort2-Key2', 2)} can be split into two txns where Cohort1 only needs to start read-only txn for its Get and Cohort2 needs to start read-write transaction for its Put.

> I guess the cohort could loop through the operations first to figure this out and then loop through them to process them.
True and valid. If we implement in this way, loop through to decide txn type sgtm. At this point we can treat the current db interface as being already stable. So you can use `Begin()` to start RW txn and `BeginReadOnly()` to start RO txn.